File: src/main/java/model/tiles/HandTiles.java
package model.tiles;

import model.basic.Tile;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class HandTiles extends Tiles {

    private Tile newTile;
    private List<Group> kong;
    private List<Group> pung;

    public HandTiles(List<Tile> tiles) {
        super(tiles);
        this.kong = new ArrayList<>();
        this.pung = new ArrayList<>();
    }

    @Override
    public void add(Tile tile) {
        if (this.newTile != null) {
            super.add(this.newTile);
        }
        this.newTile = tile;
        this.sort();
    }

    @Override
    public void remove(Tile tile) {
        this.removeForDiscard(tile);
    }

    public Tile removeForDiscard(Tile tile) {
        if (tile == null) {
            return null;
        }
        if (tile == this.newTile) {
            Tile removed = this.newTile;
            this.newTile = null;
            this.sort();
            return removed;
        }
        for (int i = 0; i < this.tiles.size(); i++) {
            Tile current = this.tiles.get(i);
            if (current.equals(tile)) {
                this.tiles.remove(i);
                this.mergeNewWithHand();
                this.sort();
                return current;
            }
        }
        if (this.newTile != null && this.newTile.equals(tile)) {
            Tile removed = this.newTile;
            this.newTile = null;
            this.sort();
            return removed;
        }
        return null;
    }

    private void mergeNewWithHand() {
        if (this.newTile != null) {
            super.add(this.newTile);
            this.newTile = null;
        }
    }

    public void addPung(Tile tile) {
        List<Tile> pung = new ArrayList<Tile>() {{
            add(new Tile(tile.getType(), tile.getNumber()));
            add(new Tile(tile.getType(), tile.getNumber()));
            add(new Tile(tile.getType(), tile.getNumber()));
        }};
        this.pung.add(new Group(pung, GroupEnum.PUNG, 0));
        this.tiles.remove(tile);
        this.tiles.remove(tile);
        this.sort();
    }

    public void addNormalKong(Tile tile) {
        List<Tile> kong = this.generateNewKong(tile);
        this.kong.add(new Group(kong, GroupEnum.NORMAL_KONG, 0));
        this.tiles.removeAll(Collections.singletonList(tile));
        this.sort();
    }

    public void addAddKong() {
        assert this.newTile != null;
        for (int i = 0; i < this.pung.size(); i++) {
            if (this.newTile.equals(this.pung.get(i).toList().get(0))) {
                this.pung.remove(i);
                break;
            }
        }
        List<Tile> kong = this.generateNewKong(this.newTile);
        this.kong.add(new Group(kong, GroupEnum.ADD_KONG, 0));
        this.newTile = null;
        this.sort();
    }

    public void addHiddenKong() {
        for (Tile tile : this.tiles) {
            if (Collections.frequency(this.tiles, tile) == 4) {
                List<Tile> kong = this.generateNewKong(tile);
                this.kong.add(new Group(kong, GroupEnum.HIDDEN_KONG, 0));
                this.tiles.removeAll(Collections.singletonList(tile));
                this.sort();
                break;
            } else if (Collections.frequency(this.tiles, tile) == 3 && tile.equals(this.newTile)) {
                List<Tile> kong = this.generateNewKong(tile);
                this.kong.add(new Group(kong, GroupEnum.HIDDEN_KONG, 0));
                this.tiles.removeAll(Collections.singletonList(tile));
                this.newTile = null;
                this.sort();
                break;
            }
        }
    }

    private List<Tile> generateNewKong(Tile tile) {
        return new ArrayList<Tile>() {{
            add(new Tile(tile.getType(), tile.getNumber()));
            add(new Tile(tile.getType(), tile.getNumber()));
            add(new Tile(tile.getType(), tile.getNumber()));
            add(new Tile(tile.getType(), tile.getNumber()));
        }};
    }

    public List<Group> getKong() {
        return kong;
    }

    public List<Group> getPung() {
        return pung;
    }

    public List<Group> getPungKong() {
        List<Group> pungKong = new ArrayList<>();
        pungKong.addAll(this.pung);
        pungKong.addAll(this.kong);
        return pungKong;
    }

    public Tile getNewTile() {
        return newTile;
    }

}

File: UnityVersion/Assets/Scripts/Core/Model/HandTiles.cs
using System.Collections.Generic;
using System.Linq;

namespace SichuanMahjong.Core.Model
{
    public class HandTiles : Tiles
    {
        private Tile newTile;
        private readonly List<Group> kong;
        private readonly List<Group> pung;

        public HandTiles(List<Tile> tiles) : base(tiles)
        {
            kong = new List<Group>();
            pung = new List<Group>();
        }

        public override void Add(Tile tile)
        {
            if (newTile != null)
            {
                base.Add(newTile);
            }
            newTile = tile;
            Sort();
        }

        public override void Remove(Tile tile)
        {
            RemoveForDiscard(tile);
        }

        public Tile RemoveForDiscard(Tile tile)
        {
            if (tile == null)
            {
                return null;
            }
            if (ReferenceEquals(tile, newTile))
            {
                Tile removed = newTile;
                newTile = null;
                Sort();
                return removed;
            }
            for (int i = 0; i < tiles.Count; i++)
            {
                Tile current = tiles[i];
                if (current.Equals(tile))
                {
                    tiles.RemoveAt(i);
                    MergeNewWithHand();
                    Sort();
                    return current;
                }
            }
            if (newTile != null && newTile.Equals(tile))
            {
                Tile removed = newTile;
                newTile = null;
                Sort();
                return removed;
            }
            return null;
        }

        private void MergeNewWithHand()
        {
            if (newTile != null)
            {
                base.Add(newTile);
                newTile = null;
            }
        }

        public void AddPung(Tile tile)
        {
            List<Tile> pungTiles = new List<Tile>
            {
                new Tile(tile.GetTileType(), tile.GetNumber()),
                new Tile(tile.GetTileType(), tile.GetNumber()),
                new Tile(tile.GetTileType(), tile.GetNumber())
            };
            pung.Add(new Group(pungTiles, GroupEnum.PUNG, 0));
            RemoveOneByValue(tile);
            RemoveOneByValue(tile);
            Sort();
        }

        public void AddNormalKong(Tile tile)
        {
            List<Tile> kongTiles = GenerateNewKong(tile);
            kong.Add(new Group(kongTiles, GroupEnum.NORMAL_KONG, 0));
            tiles.RemoveAll(t => t.Equals(tile));
            Sort();
        }

        public void AddAddKong()
        {
            for (int i = 0; i < pung.Count; i++)
            {
                if (newTile.Equals(pung[i].ToList()[0]))
                {
                    pung.RemoveAt(i);
                    break;
                }
            }
            List<Tile> kongTiles = GenerateNewKong(newTile);
            kong.Add(new Group(kongTiles, GroupEnum.ADD_KONG, 0));
            newTile = null;
            Sort();
        }

        public void AddHiddenKong()
        {
            foreach (Tile tile in tiles)
            {
                int frequency = tiles.Count(t => t.Equals(tile));
                if (frequency == 4)
                {
                    List<Tile> kongTiles = GenerateNewKong(tile);
                    kong.Add(new Group(kongTiles, GroupEnum.HIDDEN_KONG, 0));
                    tiles.RemoveAll(t => t.Equals(tile));
                    Sort();
                    break;
                }
                if (frequency == 3 && tile.Equals(newTile))
                {
                    List<Tile> kongTiles = GenerateNewKong(tile);
                    kong.Add(new Group(kongTiles, GroupEnum.HIDDEN_KONG, 0));
                    tiles.RemoveAll(t => t.Equals(tile));
                    newTile = null;
                    Sort();
                    break;
                }
            }
        }

        private static List<Tile> GenerateNewKong(Tile tile)
        {
            return new List<Tile>
            {
                new Tile(tile.GetTileType(), tile.GetNumber()),
                new Tile(tile.GetTileType(), tile.GetNumber()),
                new Tile(tile.GetTileType(), tile.GetNumber()),
                new Tile(tile.GetTileType(), tile.GetNumber())
            };
        }

        /// <summary>
        /// Java's List.remove(Object) removes the first element equal to the
        /// argument; List&lt;T&gt;.Remove uses reference semantics only when
        /// Equals is not overridden, so spell out the by-value removal.
        /// </summary>
        private void RemoveOneByValue(Tile tile)
        {
            for (int i = 0; i < tiles.Count; i++)
            {
                if (tiles[i].Equals(tile))
                {
                    tiles.RemoveAt(i);
                    return;
                }
            }
        }

        public List<Group> GetKong()
        {
            return kong;
        }

        public List<Group> GetPung()
        {
            return pung;
        }

        public List<Group> GetPungKong()
        {
            List<Group> pungKong = new List<Group>();
            pungKong.AddRange(pung);
            pungKong.AddRange(kong);
            return pungKong;
        }

        public Tile GetNewTile()
        {
            return newTile;
        }

        public override string ToString()
        {
            return "HandTiles{tiles=[" + string.Join(", ", tiles) + "], newTile=" + newTile
                   + ", pung=[" + string.Join(", ", pung) + "], kong=[" + string.Join(", ", kong) + "]}";
        }
    }
}
